doc/simd_api.rst

```diff
@@ -0,0 +1,948 @@
+   maps to the native architecture implementation for *V* and *N*, if
+   supported.
+
+   ``simd_abi::default_abi`` will use ``simd_abi::native`` if available, or
+   else fall back to the generic implementation.
+
+2. ``simd_mask<V, N, I = simd_api::default_abi>``
+
+   The result of performing a lane-wise comparison/test operation on
+   a ``simd<V, N, I>`` vector value. ``simd_mask`` objects support logical
+   operations and are used as arguments to ``where`` expressions.
+
+   ``simd_mask<V, N, I>`` is a type alias for ``simd<V, N, I>::simd_mask``.
+
+3. ``where_expression<simd<V, N, I>>``
+
+   The result of a ``where`` expression, used for masked assignment.
+
+Implementation typemaps live in the ``simd_abi`` namespace, while concrete
+implementation classes live in ``simd_detail``. A particular specialization
+for an architecture, for example 4-wide double on AVX, then requires:
+
+*  A concrete implementation class, e.g. ``simd_detail::avx_double4``.
+
+*  A specialization of its ABI map, so that ``simd_abi::avx<double, 4>::type``
+   is an alias for ``simd_detail::avx_double4``.
+
+*  A specialization of the native ABI map, so that
+   ``simd_abi::native<double, 4>::type`` is an alias for ``simd_abi::avx<double, 4>::type``.
+
+The maximum natively supported width for a scalar type *V* is recorded in
+``simd_abi::native_width<V>::value``.
+
+Class ``simd``
+^^^^^^^^^^^^^^
+
+The class ``simd<V, N, I>`` is an alias for ``simd_detail::simd_impl<I<V, N>::type>``;
+the class ``simd_detail::simd_impl<C>`` provides the public interface and
+arithmetic operators for a concrete implementation class `C`.
+
+In the following:
+
+* *S* stands for the class ``simd<V, N, I>``.
+* *s* is an SIMD value of type *S*.
+* *m* is a mask value of type ``S::simd_mask``.
+* *t*, *u* and *v* are const objects of type *S*.
+* *i* is an index of type ``int``.
+* *j* is a const object of type ``simd<U, N, J>`` where *U* is an integral type.
+* *x* is a value of type *V*.
+* *p* is a pointer to *V*.
+* *c* is a const pointer to *V* or a length *N* array of *V*.
+
+Here and below, the value in lane *i* of a SIMD vector or mask *v* is denoted by
+*v*\ `i`:sub:
+
+
+.. rubric:: Type aliases and constexpr members
+
+.. list-table::
+    :widths: 20 20 60
+    :header-rows: 1
+
+    * - Name
+      - Type
+      - Description
+
+    * - ``S::scalar_type``
+      - *V*
+      - The type of one lane of the SIMD type.
+
+    * - ``S::simd_mask``
+      - ``simd_mask<V, N, I>``
+      - The ``simd_mask`` specialization resulting from comparisons of *S* SIMD values.
+
+    * - ``S::width``
+      - ``unsigned``
+      - The SIMD width *N*.
+
+.. rubric:: Constructors
+
+.. list-table::
+    :widths: 20 80
+    :header-rows: 1
+
+    * - Expression
+      - Description
+
+    * - ``S(x)``
+      - A SIMD value *v* with *v*\ `i`:sub: equal to *x* for *i* = 0…*N*-1.
+
+    * - ``S(t)``
+      - A copy of the SIMD value *t*.
+
+    * - ``S(c)``
```

Comment:
I think that explicitly saying `S(const V*)`, or even `simd<V,N,I>(const V*)` is clearer, because that is the signature of the constructor.

Reply:
It's a matter of presenting things in terms of the methods defined in the interface code versus presenting the semantics of expressions (which in principle may be achieved with different interface code). I went with the latter, but I'm not strongly committed to this position. But again, would like to tackle this sort of finessing of the format of the documentation in a separate, later patch.